r/src/convert_array.c

```diff
@@ -100,7 +100,26 @@ static SEXP convert_array_default(SEXP array_xptr, enum VectorType vector_type,
   return result;
 }
 
-static SEXP convert_array_chr(SEXP array_xptr) {
+static SEXP convert_array_chr(SEXP array_xptr, SEXP ptype_sexp) {
+  struct ArrowSchema* schema = schema_from_array_xptr(array_xptr);
+  struct ArrowSchemaView schema_view;
+  if (ArrowSchemaViewInit(&schema_view, schema, NULL) != NANOARROW_OK) {
+    Rf_error("Invalid schema");
+  }
+
+  // If array_xptr is an extension, use default conversion
+  if (schema_view.extension_name.size_bytes > 0) {
+    if (ptype_sexp == R_NilValue) {
+      ptype_sexp = PROTECT(nanoarrow_c_infer_ptype(array_xptr_get_schema(array_xptr)));
+      SEXP default_result =
+          PROTECT(convert_array_default(array_xptr, VECTOR_TYPE_OTHER, ptype_sexp));
+      UNPROTECT(2);
+      return default_result;
+    } else {
+      return convert_array_default(array_xptr, VECTOR_TYPE_OTHER, ptype_sexp);
+    }
```

Comment:
This gave me pause. Perhaps:

```suggestion
    if (ptype_sexp == R_NilValue) {
      ptype_sexp = PROTECT(nanoarrow_c_infer_ptype(array_xptr_get_schema(array_xptr)));
    } else {
      ptype_sexp = PROTECT(ptype_sexp);
    }
    SEXP default_result =
        convert_array_default(array_xptr, VECTOR_TYPE_OTHER, ptype_sexp);
    UNPROTECT(1);
    return default_result;
```